libmemcached/tls.c

```diff
@@ -0,0 +1,595 @@
+        rc = MEMCACHED_SSL_CTX_CIPHERSUITES_LOAD_FAILED;
+        goto error;
+    }
+#endif
+
+    if (ctx_config->prefer_server_ciphers && !SSL_CTX_set_options(ssl_ctx->ctx, SSL_OP_CIPHER_SERVER_PREFERENCE)) {
+        rc = MEMCACHED_SSL_CTX_PREFER_SERVER_CIPHER_LOAD_FAILED;
+        goto error;
+    }
+
+    ptr->ssl_ctx = ssl_ctx;
+
+    return MEMCACHED_SSL_CTX_SUCCESS;
+
+    error:
+        if (ssl_ctx) {
+            _memcached_free_ssl_ctx(ptr, ssl_ctx);
+        }
+        return rc;
+}
+
+memcached_SSL_CTX *memcached_get_ssl_context_copy(const memcached_st *ptr) {
+    memcached_SSL_CTX *src_ssl_ctx;
+    memcached_SSL_CTX *dst_ssl_ctx;
+    if (ptr == NULL || ptr->ssl_ctx == NULL) {
+        return NULL;
+    }
+
+    src_ssl_ctx = (memcached_SSL_CTX*)ptr->ssl_ctx;
+
+    if (src_ssl_ctx == NULL) {
+        return NULL;
+    }
+
+    dst_ssl_ctx = (memcached_SSL_CTX *)libmemcached_calloc(ptr, 1, sizeof(memcached_SSL_CTX));
+    dst_ssl_ctx->ctx = src_ssl_ctx->ctx;
+    SSL_CTX_up_ref(src_ssl_ctx->ctx);
+
+    if (src_ssl_ctx->hostname != NULL) {
+        dst_ssl_ctx->hostname = strdup(src_ssl_ctx->hostname);
+    }
+
+    return dst_ssl_ctx;
+}
+
+memcached_return_t memcached_set_ssl_context(memcached_st *ptr, memcached_SSL_CTX *ssl_ctx) {
+  if (ptr == NULL || ssl_ctx == NULL || ssl_ctx->ctx == NULL)
+  {
+    return MEMCACHED_INVALID_ARGUMENTS;
+  }
+  ptr->ssl_ctx = ssl_ctx;
+  return MEMCACHED_SUCCESS;
+}
+
+void _memcached_free_ssl_ctx(const memcached_st *ptr, memcached_SSL_CTX *ssl_ctx)
+{
+    if (!ssl_ctx) {
+        return;
+    }
+
+    if (ssl_ctx->hostname) {
+        free(ssl_ctx->hostname);
+        ssl_ctx->hostname = NULL;
+    }
+
+    if (ssl_ctx->ctx) {
+        SSL_CTX_free(ssl_ctx->ctx);
+        ssl_ctx->ctx = NULL;
+    }
+
+    libmemcached_free(ptr, ssl_ctx);
+}
+
+
+static ssize_t handle_ssl_return_value(memcached_instance_st* instance, int rv, want_t *want){
+    if (rv <= 0) {
+        memcached_SSL *memc_ssl = (memcached_SSL*)instance->privctx;
+        int err = SSL_get_error(memc_ssl->ssl, rv);
+        switch (err) {
+            case SSL_ERROR_WANT_READ:
+                *want = WANT_READ;
+                return -1;
+            case SSL_ERROR_WANT_WRITE:
+                *want = WANT_WRITE;
+                return -1;
+            case SSL_ERROR_ZERO_RETURN:
+                // The TLS peer has closed the connection
+                memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT,
+                memcached_literal_param(ERR_reason_error_string(ERR_get_error())));
+                return 0;
+            default:
+                memcached_set_error(*instance, MEMCACHED_TLS_ERROR, MEMCACHED_AT,
+                                memcached_literal_param(ERR_reason_error_string(ERR_get_error())));
+                return -1;
+        }
+    }
+    return rv;
+}
+
+static bool handle_renegotation_error(memcached_instance_st* instance, want_t *want) {
```

Comment:
Typo: `renegotiation`.

It is typically not recommended to support TLS re-negotiation at least from the server side because re-negotiation of TLS is very complicated, and it can lead to security issues such as DDoS attack and man-in-the-middle attacks. Not sure if we need to handle re-negotiation error from the client side. It is probably better to disable this and close the connection always.

Reply:
The function name isn't right and should be updated to handle_want_error(), since it actually handles openssl's WANT_READ/WANT_WRITE errors that can happen from protocol reasons (e.g. renegotiation, key update) or with a non blocking socket if not enough data was available at that time to complete the operation. 

From here: https://github.com/openssl/openssl/issues/11211#event-3094172350
In TLS1.3, SSL_write can yield SSL_ERROR_WANT_READ, and SSL_read also can yield SSL_ERROR_WANT_WRITE

In the OSS's Redis they do handle WANT errors, same as in hiredis, but it is possible to close the connection each time we get WANT error. WDYT?

Reply:
OK, please rename this method. Don't call it `re-negotiation`

Reply:
changed